sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/Hotspots/ControllingPermissionsBase.cs

```diff
@@ -0,0 +1,78 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class ControllingPermissionsBase<TSyntaxKind> : SonarDiagnosticAnalyzer
+        where TSyntaxKind : struct
+    {
+        protected const string DiagnosticId = "S4834";
+        protected const string MessageFormat = "Make sure that permissions are controlled safely here.";
+
+        protected ObjectCreationTracker<TSyntaxKind> ObjectCreationTracker { get; set; }
+        protected InvocationTracker<TSyntaxKind> InvocationTracker { get; set; }
+        protected PropertyAccessTracker<TSyntaxKind> PropertyAccessTracker { get; set; }
+        protected MethodDeclarationTracker<TSyntaxKind> MethodDeclarationTracker { get; set; }
+        protected BaseTypeTracker<TSyntaxKind> BaseTypeTracker { get; set; }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            ObjectCreationTracker.Track(context,
+                ObjectCreationTracker.MatchConstructors(
+                    KnownType.System_Security_Permissions_PrincipalPermission));
+
+            ObjectCreationTracker.Track(context,
+                ObjectCreationTracker.MatchAnyTypeThatImplements(
+                    KnownType.System_Security_Principal_IIdentity,
+                    KnownType.System_Security_Principal_IPrincipal));
+
+            InvocationTracker.Track(context,
+                InvocationTracker.MatchSimpleNames(
+                    new MethodSignature(KnownType.System_Security_Principal_WindowsIdentity, "GetCurrent"),
+                    new MethodSignature(KnownType.System_IdentityModel_Tokens_SecurityTokenHandler, "ValidateToken"),
+                    new MethodSignature(KnownType.System_AppDomain, "SetPrincipalPolicy"),
+                    new MethodSignature(KnownType.System_AppDomain, "SetThreadPrincipal")));
+
+            PropertyAccessTracker.Track(context,
+                PropertyAccessTracker.MatchSimpleNames(
+                    new MethodSignature(KnownType.System_Web_HttpContext, "User"),
+                    new MethodSignature(KnownType.System_Threading_Thread, "CurrentPrincipal")));
+
+            MethodDeclarationTracker.Track(context,
+                MethodDeclarationTracker.AnyParameterIsOfType(
+                    KnownType.System_Security_Principal_IIdentity,
+                    KnownType.System_Security_Principal_IPrincipal));
+
+            MethodDeclarationTracker.Track(context,
+                MethodDeclarationTracker.DecoratedWithAnyAttribute(
+                    KnownType.System_Security_Permissions_PrincipalPermissionAttribute));
+
+            BaseTypeTracker.Track(context,
+                BaseTypeTracker.WhenDerivesFrom(
+                    KnownType.System_Security_Principal_IIdentity));
+
+            BaseTypeTracker.Track(context,
```

Comment:
There is now an overload of _BaseTypeTracker.WhenDerivesFrom_ that accepts a list of types (although it might be better if it was renamed to _WhenDerivesFromAny_).